src/plugins/data/server/search/session/session_service.ts

```diff
@@ -6,47 +6,549 @@
-    return () => ({
-      getId: () => {
-        throw new Error('getId not implemented in OSS search session service');
-      },
-      trackId: async () => {},
-      getSearchIdMapping: async () => new Map<string, string>(),
-      save: async () => {
-        throw new Error('save not implemented in OSS search session service');
-      },
-      get: async () => {
-        throw new Error('get not implemented in OSS search session service');
-      },
-      find: async () => {
-        throw new Error('find not implemented in OSS search session service');
-      },
-      update: async () => {
-        throw new Error('update not implemented in OSS search session service');
-      },
-      extend: async () => {
-        throw new Error('extend not implemented in OSS search session service');
-      },
-      cancel: async () => {
-        throw new Error('cancel not implemented in OSS search session service');
-      },
-      delete: async () => {
-        throw new Error('delete not implemented in OSS search session service');
-      },
-      getConfig: () => {
-        return null;
+interface StartDependencies {
+  taskManager: TaskManagerStartContract;
+}
+
+const DEBOUNCE_UPDATE_OR_CREATE_WAIT = 1000;
+const DEBOUNCE_UPDATE_OR_CREATE_MAX_WAIT = 5000;
+
+interface UpdateOrCreateQueueEntry {
+  deps: SearchSessionDependencies;
+  user: AuthenticatedUser | null;
+  sessionId: string;
+  attributes: Partial<SearchSessionSavedObjectAttributes>;
+  resolve: () => void;
+  reject: (reason?: unknown) => void;
+}
+
+function sleep(ms: number) {
+  return new Promise((r) => setTimeout(r, ms));
+}
+export class SearchSessionService
+  implements ISearchSessionService<SearchSessionSavedObjectAttributes>
+{
+  private sessionConfig: SearchSessionsConfigSchema;
+  private readonly updateOrCreateBatchQueue: UpdateOrCreateQueueEntry[] = [];
+  private security?: SecurityPluginSetup;
+  private setupCompleted = false;
+
+  constructor(
+    private readonly logger: Logger,
+    private readonly config: ConfigSchema,
+    private readonly version: string
+  ) {
+    this.sessionConfig = this.config.search.sessions;
+  }
+
+  public setup(core: CoreSetup, deps: SetupDependencies) {
+    this.security = deps.security;
+    const taskDeps = {
+      config: this.config,
+      taskManager: deps.taskManager,
+      logger: this.logger,
+    };
+
+    registerSearchSessionsTask(
+      core,
+      taskDeps,
+      SEARCH_SESSIONS_TASK_TYPE,
+      'persisted session progress',
+      checkPersistedSessionsProgress
+    );
+
+    registerSearchSessionsTask(
+      core,
+      taskDeps,
+      SEARCH_SESSIONS_CLEANUP_TASK_TYPE,
+      'non persisted session cleanup',
+      checkNonPersistedSessions
+    );
+
+    registerSearchSessionsTask(
+      core,
+      taskDeps,
+      SEARCH_SESSIONS_EXPIRE_TASK_TYPE,
+      'complete session expiration',
+      checkPersistedCompletedSessionExpiration
+    );
+
+    this.setupCompleted = true;
```

Comment:
Mostly a precaution, because task_manager plugin is optional and it is missing in functional non-xpack tests. So we have to have a code branch now where search_service doesn't have task_manager and it doesn't initialize session_service: 

https://github.com/elastic/kibana/blob/1bdbafef7cc1e458e785f0cc8d7d823f3cf7f577/src/plugins/data/server/search/search_service.ts#L159-L165


This makes it possible so that setup is never called and I wanted a clear error in case start is called by mistake in this case